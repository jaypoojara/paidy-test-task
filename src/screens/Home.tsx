import React, { useEffect, useRef, useState } from "react";
import {
  View,
  Text,
  SafeAreaView,
  StatusBar,
  FlatList,
  TextInput,
  StyleSheet,
} from "react-native";
import AsyncStorage from "@react-native-async-storage/async-storage";
import { ITodoItem } from "../types/interface";
import TodoItem from "../components/TodoItem";
import Button from "../components/Button";
import { colors } from "../types/colors";
import { asyncStoragekeys } from "../types/asyncStoragekeys";

const Home = () => {
  // Ref to FlatList for scrolling
  const flatListRef = useRef(null);

  // State for input text, selected todo index, and todos list
  const [text, setText] = useState<string>("");
  const [selectedTodoIndex, setSelectedTodoIndex] = useState<number | null>(
    null
  );
  const [todos, setTodos] = useState<ITodoItem[]>([]);

  useEffect(() => {
    // Load todos from AsyncStorage on component mount
    getTodos();
  }, []);

  // Retrieve todos from AsyncStorage
  const getTodos = async () => {
    try {
      const value = await AsyncStorage.getItem(asyncStoragekeys.Todos);
      if (value !== null) {
        setTodos(JSON.parse(value));
      }
    } catch (e) {}
  };

  // Add or update a todo item
  const onClickAddUpdate = () => {
    const updateTodos = todos;
    if (selectedTodoIndex !== null) {
      // Update existing todo
      updateTodos[selectedTodoIndex].title = text;
    } else {
      // Add new todo
      updateTodos.push({ title: text });
    }
    storeTodos(updateTodos);
    setTodos(updateTodos);
    setText("");
    setSelectedTodoIndex(null);

    // Scroll to the end of the list after updating
    setTimeout(() => {
      flatListRef.current.scrollToEnd();
    }, 100);
  };

  // Remove a todo item
  const onClickRemove = (index: number) => {
    const updateTodos = todos;
    updateTodos.splice(index, 1);
    storeTodos(updateTodos);
    setTodos([...updateTodos]);
    setSelectedTodoIndex(null);
    setText("");
  };

  // Click handler for selecting a todo
  const onClickTodo = (index: number) => {
    setText(todos[index].title);
    setSelectedTodoIndex(index);
  };

  // Store todos to AsyncStorage
  const storeTodos = async (todos: ITodoItem[]) => {
    try {
      await AsyncStorage.setItem(asyncStoragekeys.Todos, JSON.stringify(todos));
    } catch (e) {}
  };

  // Render each todo item
  const renderItem = ({ item, index }: { item: ITodoItem; index: number }) => (
    <TodoItem
      title={item.title}
      index={index}
      onPressRemove={onClickRemove}
      onPressTodo={onClickTodo}
    />
  );

  return (
    <SafeAreaView style={styles.container}>
      {/* Header */}
      <Text style={styles.headerText}>TODO:</Text>

      {/* List of todos */}
      <FlatList
        ref={flatListRef}
        showsVerticalScrollIndicator={false}
        contentContainerStyle={styles.flatListContent}
        data={todos}
        keyExtractor={(_, index) => index.toString()}
        renderItem={renderItem}
        ListEmptyComponent={
          // Empty container
          <View style={styles.emptyContainer}>
            <Text style={styles.emptyText}>Todos not found!</Text>
          </View>
        }
      />

      {/* Input container */}
      <View style={styles.inputContainer}>
        <TextInput
          style={styles.input}
          placeholder="Enter here"
          value={text}
          onChangeText={setText}
        />
        {/* Button to add or update a todo */}
        <Button
          onPress={() => onClickAddUpdate()}
          title={selectedTodoIndex != null ? "UPDATE" : "ADD"}
          radius={15}
        />
      </View>
    </SafeAreaView>
  );
};

export default Home;

// Styles for the screen
const styles = StyleSheet.create({
  container: {
    paddingTop: StatusBar.currentHeight,
    marginHorizontal: 15,
    flex: 1,
    paddingBottom: 20,
  },
  headerText: {
    fontSize: 20,
    fontWeight: "bold",
    color: colors.primary,
    marginBottom: 10,
  },
  flatListContent: {
    flexGrow: 1,
  },
  inputContainer: {
    marginVertical: 5,
    paddingVertical: 10,
    paddingHorizontal: 15,
    backgroundColor: colors.white,
    flexDirection: "row",
    borderRadius: 15,
  },
  input: {
    flex: 1,
    fontWeight: "500",
  },
  emptyContainer: {
    flex: 1,
    justifyContent: "center",
    alignItems: "center",
  },
  emptyText: {
    fontSize: 16,
    fontWeight: "bold",
    color: colors.lable,
  },
});
