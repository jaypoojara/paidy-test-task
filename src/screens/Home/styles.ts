import { StatusBar, StyleSheet } from "react-native";
import { Colors } from "../../types/colors";

// Styles for the screen
const styles = StyleSheet.create({
  container: {
    paddingTop: StatusBar.currentHeight,
    marginHorizontal: 15,
    flex: 1,
    paddingBottom: 20,
  },
  headerText: {
    fontSize: 20,
    fontWeight: "bold",
    color: Colors.primary,
    marginBottom: 10,
  },
  flatListContent: {
    flexGrow: 1,
  },
  inputContainer: {
    marginVertical: 5,
    paddingVertical: 10,
    paddingHorizontal: 15,
    backgroundColor: Colors.white,
    flexDirection: "row",
    borderRadius: 15,
  },
  input: {
    flex: 1,
    fontWeight: "500",
  },
});

export { styles };
