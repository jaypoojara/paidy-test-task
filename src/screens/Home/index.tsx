import React, { useEffect, useRef, useState } from "react";
import { View, Text, SafeAreaView, FlatList, TextInput } from "react-native";
import AsyncStorage from "@react-native-async-storage/async-storage";
import { IToDoItem } from "../../types/interface";
import ToDoItem from "../../components/ToDoItem";
import Button from "../../components/Button";
import { asyncStoragekeys } from "../../types/asyncStoragekeys";
import ListEmpt from "../../components/ListEmpt";
import { Styles } from "./styles";

const Home = () => {
  // Ref to FlatList for scrolling
  const flatListRef = useRef(null);

  // State for input text, selected todo index, and todos list
  const [text, setText] = useState<string>("");
  const [selectedTodoIndex, setSelectedTodoIndex] = useState<number | null>(
    null
  );
  const [todos, setTodos] = useState<IToDoItem[]>([]);

  useEffect(() => {
    // Load todos from AsyncStorage on component mount
    getTodos();
  }, []);

  // Retrieve todos from AsyncStorage
  const getTodos = async () => {
    try {
      const value = await AsyncStorage.getItem(asyncStoragekeys.Todos);
      if (value !== null) {
        setTodos(JSON.parse(value));
      }
    } catch (e) {}
  };

  // Add or update a todo item
  const onClickAddUpdate = () => {
    const updateTodos = todos;
    if (selectedTodoIndex !== null) {
      // Update existing todo
      updateTodos[selectedTodoIndex].title = text;
    } else {
      // Add new todo
      updateTodos.push({ title: text });
    }
    storeTodos(updateTodos);
    setTodos(updateTodos);
    setText("");
    setSelectedTodoIndex(null);

    // Scroll to the end of the list after updating
    setTimeout(() => {
      flatListRef.current.scrollToEnd();
    }, 100);
  };

  // Remove a todo item
  const onClickRemove = (index: number) => {
    const updateTodos = todos;
    updateTodos.splice(index, 1);
    storeTodos(updateTodos);
    setTodos([...updateTodos]);
    setSelectedTodoIndex(null);
    setText("");
  };

  // Click handler for selecting a todo
  const onClickTodo = (index: number) => {
    setText(todos[index].title);
    setSelectedTodoIndex(index);
  };

  // Store todos to AsyncStorage
  const storeTodos = async (todos: IToDoItem[]) => {
    try {
      await AsyncStorage.setItem(asyncStoragekeys.Todos, JSON.stringify(todos));
    } catch (e) {}
  };

  // Render each todo item
  const renderItem = ({ item, index }: { item: IToDoItem; index: number }) => (
    <ToDoItem
      title={item.title}
      index={index}
      onPressRemove={onClickRemove}
      onPressTodo={onClickTodo}
    />
  );

  return (
    <SafeAreaView style={Styles.container}>
      {/* Header */}
      <Text style={Styles.headerText}>TODO:</Text>

      {/* List of todos */}
      <FlatList
        ref={flatListRef}
        showsVerticalScrollIndicator={false}
        contentContainerStyle={Styles.flatListContent}
        data={todos}
        keyExtractor={(_, index) => index.toString()}
        renderItem={renderItem}
        ListEmptyComponent={<ListEmpt />}
      />

      {/* Input container */}
      <View style={Styles.inputContainer}>
        <TextInput
          style={Styles.input}
          placeholder="Enter here"
          value={text}
          onChangeText={setText}
        />
        {/* Button to add or update a todo */}
        <Button
          onPress={() => onClickAddUpdate()}
          title={selectedTodoIndex != null ? "UPDATE" : "ADD"}
          radius={15}
        />
      </View>
    </SafeAreaView>
  );
};

export default Home;
