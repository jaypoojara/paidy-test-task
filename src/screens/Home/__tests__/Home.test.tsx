import React from "react";
import { render, fireEvent, waitFor } from "@testing-library/react-native";
import AsyncStorage from "@react-native-async-storage/async-storage";
import Home from "../index";

// Mock AsyncStorage methods
jest.mock("@react-native-async-storage/async-storage", () => ({
  getItem: jest.fn(),
  setItem: jest.fn(),
}));

describe("Home Component", () => {
  it("renders correctly", () => {
    const element = render(<Home />);
    expect(element).toMatchSnapshot();
  });

  it("renders correctly and adds a new todo", async () => {
    AsyncStorage.getItem.mockResolvedValue(JSON.stringify([]));
    AsyncStorage.setItem.mockResolvedValue(null);

    const { getByPlaceholderText, getByText } = render(<Home />);

    const inputElement = getByPlaceholderText("Enter here");
    expect(inputElement).toBeDefined();

    fireEvent.changeText(inputElement, "New Todo");
    const addButton = getByText("ADD");
    fireEvent.press(addButton);

    await waitFor(() => expect(AsyncStorage.setItem).toHaveBeenCalled());
  });
});
