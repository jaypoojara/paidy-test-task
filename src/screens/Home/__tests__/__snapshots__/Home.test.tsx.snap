// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Component renders correctly 1`] = `
<RCTSafeAreaView
  style={
    {
      "flex": 1,
      "marginHorizontal": 15,
      "paddingTop": null,
    }
  }
>
  <Text
    style={
      {
        "color": "#034096",
        "fontSize": 20,
        "fontWeight": "bold",
        "marginBottom": 10,
      }
    }
  >
    TODO:
  </Text>
  <RCTScrollView
    contentContainerStyle={
      {
        "flexGrow": 1,
      }
    }
    data={[]}
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    showsVerticalScrollIndicator={false}
    stickyHeaderIndices={[]}
    viewabilityConfigCallbackPairs={[]}
  >
    <View />
  </RCTScrollView>
  <View
    style={
      {
        "backgroundColor": "#FFFFFF",
        "borderRadius": 15,
        "flexDirection": "row",
        "marginVertical": 5,
        "paddingHorizontal": 15,
        "paddingVertical": 10,
      }
    }
  >
    <TextInput
      onChangeText={[Function]}
      placeholder="Enter here"
      style={
        {
          "flex": 1,
          "fontWeight": "500",
        }
      }
      value=""
    />
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "backgroundColor": "#034096",
          "borderRadius": 15,
          "opacity": 1,
          "paddingHorizontal": 30,
          "paddingVertical": 12,
        }
      }
      testID="buttonTestID"
    >
      <Text
        style={
          {
            "color": "#FFFFFF",
            "fontSize": 18,
            "fontWeight": "bold",
          }
        }
      >
        ADD
      </Text>
    </View>
  </View>
</RCTSafeAreaView>
`;
